### PR TITLE
Stochastic FW speed up

### PR DESCRIPTION
**Pull request by GeoffNN**
Bottlenecks are due to limitations of `scipy.sparse.csr_matrix`. These objects are not optimized to be indexed before vector-matrix or matrix-vector multiplications.

### CODE REVIEW

File: copt/utils.py [hunk 1]
```diff
@@ -77,6 +77,45 @@ def safe_sparse_add(a, b):
...
+    res = np.zeros(d)
+    assert x.shape[0] == len(idx)
+    for k, i in np.ndenumerate(idx):
+        for j in range(indptr[i], indptr[i+1]):
+            j_idx = indices[j]
+            res[j_idx] += x[k] * data[j]
+    return res
+
+
+@njit(nogil=True)
```
Comment (fabianp) [inline copt/utils.py:101]: Are both functions needed? It looks to me that one is a more general function that the other

File: copt/utils.py [hunk 2]
```diff
@@ -77,6 +77,45 @@ def safe_sparse_add(a, b):
...
+def fast_csr_mv(data, indptr, indices, x, idx):
+    """
```
Comment (fabianp) [inline copt/utils.py:103]: please add description of what the function does

File: copt/utils.py (hunk 1 above)
Comment (GeoffNN) [reply copt/utils.py:101]: I think they are: they're both matrix-vector multiplications, but one is vM, the other is Mv.

**Comment by GeoffNN** [comment]
Changing to epochs seems to introduce weird behavior. It would be great to have someone else check it out. @fabianp @gideonite maybe?

File: copt/utils.py (hunk 2 above)
Comment (GeoffNN) [reply copt/utils.py:106]: Done

File: copt/utils.py (hunk 1 above)
Comment (GeoffNN) [reply copt/utils.py:104]: If we double-transpose, we get `vM = (M^Tv^T)^T`, but then `v^T` is a row, and not a column.

File: copt/utils.py [hunk 3]
```diff
@@ -77,6 +77,45 @@ def safe_sparse_add(a, b):
...
 
+@njit(nogil=True)
+def fast_csr_vm(x, data, indptr, indices, d, idx):
+    """
+
+    :param x:
+    :param data:
+    :param indptr:
+    :param indices:
+    :param d:
```
[the remark this answers is not in the document]
Comment (GeoffNN) [reply copt/utils.py]: Done

**Comment by fabianp** [comment]
i'd like to merge this since it's a clear improvement wrt to the previous code. We can continue improving this after merging. Are you OK with this @GeoffNN ?

**Comment by fabianp** [comment]
Also, when you merge please also consider merging #53

**Comment by GeoffNN** [comment]
Merging.
Comment as a todo : benchmark differences between `numpy.random.choice` and `numpy.random.randint` if we're sampling with replacement.

**Pull request merged**